### DISCUSSION TITLE
How to use released docker images?

### DISCUSSION THREAD

**Discussion by olorin37** [tweag/nickel#1950] [Q&A]
I afraid that something is wrong with provided docker images:

After importing image to podman with:
```
podman import nickel-x86_64-docker-image.tar.gz nickel:1.6.0
```
A cannot start anything by:
```
❯ podman run -it --rm nickel:1.6.0
Error: no command or entrypoint provided, and no CMD or ENTRYPOINT from image

❯ podman run -it --rm nickel:1.6.0 nickel
Error: runc: runc create failed: unable to start container process: exec: "nickel": executable file not found in $PATH: OCI runtime attempted to invoke a command that was not found

❯ podman run -it --rm nickel:1.6.0 /bin/nickel
Error: runc: runc create failed: unable to start container process: exec: "/bin/nickel": stat /bin/nickel: no such file or directory: OCI runtime attempted to invoke a command that was not found
```

But when I inspect the structure of the image from other container 
```
❯ podman run -it --rm --mount type=image,source=nickel:1.6.0,destination=/image,rw=true  an-image-with-bash bash
root@07bc408667a7:/# ls /image/
0ab957d91868fcef2f474cba47cf91ed4f793f7695f496b93dca4600c0670cd3
0ce4e59ceffc218886bcfbb9d112413d9e168a69887bcdeed68adca3b88efcbf.json
1320c3a1593dbc6a719e76793fdf53eb4af2225a8eefc8a06504f3e5395ed03e
manifest.json
```
The same effect I can achieve using `COPY --from=nickel:1.6.0 / /image` to build another container.

Is something wrong with docker image distribution or I should import it in different way?

**Comment by yannham** [comment]
Hi, I personally don't use podman, but I just tested the latest image wit the following commands:

```console
$ sudo docker load < ~/nickel-x86_64-docker-image.tar.gz
$ sudo docker run --rm -it nickel:1.6.0 --version
nickel-lang-cli nickel 1.6.0 (rev 3441781)
```

which seems to work.

**Reply by yannham** [reply]
Maybe you haven't load the image properly before?

**Reply by olorin37** [reply]
OK, so it seems that it is docker image format not OCI. I'll try with docker and probably podman has some docker-compatible option... I'll post the results here.

**Reply by olorin37** [reply]
OK I know what happened. I just mistaken `load` with `import`. Both do something similar, but different:
- docker/podman import creates an image with rootfs created based on files supplied by tar archive which is provided (rootfs is the same as contents of tar)
- docker/podman load does what I meant, which is registration of an image provided in docker or OCI format which _accidentally_ is also a tar but with specific internal structure.

Sorry for noise in discussions, my fault, forgot that `import` command exists.  _Sadly I still cannot load the image to the podman local registry, I'll update this discussion with solution when I get there_

**Reply by olorin37** [reply]
I found a moment for installing `docker` - _many had been changed from the time when I used it last time (versions around 20, rootless and VMs)_

My result is this:
```
gramsz@lorien ~/Downloads> docker load -i nickel-x86_64-docker-image.tar.gz
b63cbdd55804: Loading layer [==================================================>]  20.13MB/20.13MB
434327c6ef18: Loading layer [==================================================>]  10.24kB/10.24kB
Loaded image: nickel:1.6.0
gramsz@lorien ~/Downloads> podman load -i nickel-x86_64-docker-image.tar.gz
Error: payload does not match any of the supported image formats:
 * oci: open nickel-x86_64-docker-image.tar.gz/index.json: not a directory
 * oci-archive: loading index: open /var/tmp/container_images_oci1543564962/index.json: no such file or directory
 * docker-archive: parsing image configuration: json: cannot unmarshal string into Go struct field ImageConfig.config.Entrypoint of type []string
 * dir: open nickel-x86_64-docker-image.tar.gz/manifest.json: not a directory
```

I'll investigate why this doesn't work for podman. I remember this was how I was moving images between docker and podman
```
docker save | podman load # and vice versa:
podman save | docker load
```
So, surely it works and also can work right now...

**Reply by olorin37** [reply]
What a moment I just fetched `https://github.com/tweag/nickel/releases/download/1.7.0/nickel-x86_64-docker-image.tar.gz` and got tag `nickel:1.6.0`? 

This is strange:
```
gramsz@lorien ~/Downloads [2]> docker run -it --rm nickel:1.6.0  --version
nickel-lang-cli nickel 1.6.0 (rev 95967eb)
```

![image](https://github.com/tweag/nickel/assets/5931253/9f23344c-0037-4392-a275-ae4867c2937d)

OK, so for creating the docker image commit 1.7.0 - 2 is taken ... Probably it should be considered as bug.. What do you think?

**Reply by yannham** [reply]
Hmm yeah, it's strange. Beside the version everything else should be at feature parity - this is the last commit before the version bump - but it's still a bug. @vkleen do you know what could've happened here?

**Reply by vkleen** [reply]
Hmm, the GitHub workflow generates the docker image tag out of `nix eval --raw .#dockerImage.imageTag` which is read from Cargo.toml. Does the release script update that before creating the release tag and triggering the workflow?

**Reply by yannham** [reply]
> Hmm, the GitHub workflow generates the docker image tag out of nix eval --raw .#dockerImage.imageTag which is read from Cargo.toml. Does the release script update that before creating the release tag and triggering the workflow?

But from which branch, `master` ? If this is the case, no, usually master is updated last, backporting the version number. after the fact. The right thing to do is probably to target `stable` instead, which at the time of the release, should always have the exact latest release. Notwithstanding the current workflow (maybe we should bump the version first in master, with an alpha or nightly or whatever, There could also be a race condition where master get more commits in the meantime; we could have a `stable` branch ready for a few days and only release later, for example. So the safest is to build the docker image from stable, I believe.

**Reply by olorin37** [reply]
I just created an issue for the versioning problem #1956  as I also encounter the problem for binary release, it returns 1.6.0 version for `nickel --version` and points to the same commit 95967eb.
